### ISSUE TITLE
Provide a portable version

### ISSUE THREAD

**Issue by insanetesterftw** [dnGrep/dnGrep#238]
Provide a version that doesn't require installation.

Thank you.

**Comment by JVimes** [comment]
Good idea. There used to be one. I think it got lost somewhere along the way when moving the build to Appveyor.

**Comment by smaragdus** [comment]
Beside no installation I hope that data and settings will be saved in program folder (no writing to AppData and Windows registry).

**Comment by vatterspun** [comment]
Portability can be enabled using a program called Universal Extractor, which unpacks the installer and does not require admin rights.  The program does still write some window settings to the _C:\Users\USER\AppData\Local\dnGrep_ directory, which is not ideal but are only minor window settings.  This is discussed here: https://www.portablefreeware.com/forums/viewtopic.php?p=94458#p94458

A version of the program that writes ALL settings to the same folder would be helpful because it makes backing up or syncronizing programs between computers easier.  I also sometimes run multiple versions of the same program with slightly different settings to speed up my processes.  For example, one copy of dngrep could have a unique starting folder, all plugins turned off, set to ignore archives, different bookmarks, and even use a different skin.

I currently have this enabled on my system with two separate folders with dnGrep in each:

![image](https://user-images.githubusercontent.com/1289134/87261148-204e8c80-c483-11ea-9f4e-6ac8edf7d0c2.png)

---

One negative is that this can make updating the client more complicated as you have to know to update some files and not others, so it is recommended as an optional setting or separate version.

**Comment by RoyiAvital** [comment]
Having portable mode in a ZIP file would be great.

@vatterspun , I think the `%APPDATA%` can be overwritten by a launcher.
Unless the application gets it not by the ENV variable.

**Comment by vatterspun** [comment]
> Having portable mode in a ZIP file would be great.

👍 

> @vatterspun , I think the `%APPDATA%` can be overwritten by a launcher.
> Unless the application gets it not by the ENV variable.

It depends -- I've messed around with this in a variety of programs and some software just ignores your ENV variables and does a hard link to the appdata folder.  As a result and for ease of backup/sync/etc., I recommend keeping any files and settings all in the local folder.

**Comment by doug24** [comment]
Made changes in [2.9.291](https://github.com/dnGrep/dnGrep/releases/tag/v2.9.291.0) so no files are loaded or written outside of the local directory when the local directory is writable.

**Comment by RoyiAvital** [comment]
@doug24 , This is great!
Any chance to also have a run flag to set to portable mode? So in case it is not in a writeable folder and error will be shown?

**Comment by doug24** [comment]
I don't think this is necessary. If you can create or copy files into the folder - like you did when you extracted or copied the dnGrep executable files - then you are good to go.    If in doubt, run dnGrep once and check for a new or updated dnGrep.settings.dat file.

**Comment by sergeevabc** [comment]
Each time I pass by this app (since in the 21 century we still have no consensus on what GREP app is perfect), it amazes me by neglecting basic things like lack of portable ZIP (yet having, ROFL, “[startup accelerator][1]”) and lack of Windows theme support (yet having 3 palettes). Luckily, the first issue could be mitigated by [lessmsi][2] x dnGREP.bla-bla.msi -> [dnGREP_29326x64.zip](https://github.com/dnGrep/dnGrep/files/6246317/dnGREP_29326x64.zip)

![](https://i3.imageban.ru/out/2021/04/01/aa0d35016c8a867c4a998127c5c32d36.png)

[1]: https://github.com/dnGrep/dnGrep/issues/143#issuecomment-267771922
[2]: https://github.com/activescott/lessmsi/releases

**Comment by doug24** [comment]
I really don't get the point of your post.  If you or anyone else wants to contribute to the project, you are more than welcome.  I don't own the project, I do this on my own time and get nothing in return.  I just started taking on tasks because I didn't want to see the project die.  I prioritize what I do based on what I think will benefit the most people, and also what I'm interested in.

Does anyone want to make a script for AppVeyor to create a zip file from the build?  Great, do it.

dnGrep doesn't have Windows Theme support because Microsoft never put it in WPF.  Adding the themes was the next best thing, but it was a hell of a lot of work.  Maybe you'd like to see dnGrep rewritten using UWP or WinUI?  The startup acceleration is something the original author of dnGrep put in ages ago - I've thought of taking it out, but never spent the time.

**Comment by doug24** [comment]
FYI - I have a portable zip file package ready for the next release.

**Comment by doug24** [comment]
Now out in v3.2.242

**Comment by RoyiAvital** [comment]
Where are the settings kept?
The use case is upgrading portable version.
Which file should I retain to have the settings move to the upgraded version?

**Comment by doug24** [comment]
The configuration data is these files and directories:
- dnGREP.Settings.dat
- dnGrep.keyboard.xml
- bookmarks.xml
- Scripts folder
- Filters folder

If you customized any themes, they will be in `.xaml` files.
If you have customized any syntax highlighting, they will be in `.xshd` files.
If you have set up any custom plugins, that will be under the Plugins directory.

You can move all these files (except plugins) to a location of your choosing in the Options dialog. [Read more here](https://github.com/dnGrep/dnGrep/wiki/Directories). If you choose to move either directory, dnGrep will create a file "dnGrep.config.xml" in the default directory that contains the paths to the new directories, so don't delete that file when upgrading.

**Comment by RoyiAvital** [comment]
It would be great to have all those in a single folder in Portable Mode.
That way user needs only to delete all files but those and then unzip the ZIP file.

**Comment by doug24** [comment]
You can do that now. Just set the configuration directory to wherever you want in the Options dialog.  The application will move the files for you. Just don't delete the dnGrep.config.xml file.

<img width="600" alt="Image" src="https://github.com/user-attachments/assets/9cb841a5-c086-43c0-9a4e-a77849342850" />

**Comment by RoyiAvital** [comment]
I meant the default in **the Portable case** should be `dbGrep\Settings` so when upgrading all but that folder is deleted.
